### PR TITLE
[SPARK-21592][BUILD] Skip maven-compiler-plugin main and test compilations in Maven build

### PR DESCRIPTION
**Pull request by gslowikowski**
`scala-maven-plugin` in `incremental` mode compiles `Scala` and `Java` classes. There is no need to execute `maven-compiler-plugin` goals to compile (in fact recompile) `Java`.

This change reduces compilation time (over 10% on my machine).

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -2022,14 +2020,8 @@
...
           <configuration>
-            <source>${java.version}</source>
-            <target>${java.version}</target>
-            <encoding>UTF-8</encoding>
-            <maxmem>1024m</maxmem>
-            <fork>true</fork>
-            <compilerArgs>
-              <arg>-Xlint:all,-serial,-path</arg>
-            </compilerArgs>
+            <skipMain>true</skipMain> <!-- skip compile -->
```
Comment (srowen) [inline pom.xml]: I think you may be right. However if so you can remove the configuration of `maven-compiler-plugin` in the two submodules that further refine the config.

File: pom.xml [hunk 2]
```diff
@@ -1972,14 +1972,12 @@
...
             <execution>
               <id>scala-compile-first</id>
-              <phase>process-resources</phase>
               <goals>
                 <goal>compile</goal>
               </goals>
             </execution>
             <execution>
               <id>scala-test-compile-first</id>
-              <phase>process-test-resources</phase>
```
Comment (srowen) [inline pom.xml]: Why is this necessary/OK?

File: pom.xml (hunk 1 above)
Comment (gslowikowski) [inline pom.xml]: Yes, I forgot about it.

File: pom.xml (hunk 2 above)
Comment (gslowikowski) [inline pom.xml]: Changing phase to one before `compile` is very old trick, used when `scala-maven-plugin` compiles only Scala classes and this compilation must be performed before Java classes compilation (by `maven-compiler-plugin`) because Java classes depend on Scala classes.
In incremental compilation mode SBT compiler is used internally and it compiles everything. Because `maven-compiler-plugin` is disabled (compilations are skipped), there is no need to change default phase.

**Comment by gslowikowski** [comment]
Do you want me to update pull?

**Comment by srowen** [comment]
Yeah then I can run tests

**Comment by gslowikowski** [comment]
Updated.

**Comment by srowen** [comment]
Likewise @vanzin any thoughts on this one? because it touches the compilation and build

**Comment by srowen** [comment]
This will need a JIRA @gslowikowski , by the way

**Comment by srowen** [comment]
@gslowikowski I'll merge this if you'll make a JIRA and link it here via the title

**Comment by gslowikowski** [comment]
Created https://issues.apache.org/jira/browse/SPARK-21592

**Comment by srowen** [comment]
Looks good @gslowikowski , just put `[SPARK-21592]` at the start of the title to link them.

**Comment by srowen** [comment]
Merged to master. I'll watch the builds to make sure it's really happy with this change.
